### ISSUE TITLE
Problems with valet on fresh install of macOS Sierra, PHP7.1 Homebrew and Valet 2.0.3

### ISSUE THREAD

**Issue by indianafollonier** [laravel/valet#299]
Hi,

I am having an issue with getting valet to work.

I just went through and did a clean install of Sierra, Homebrew, PHP7.1 and composer (and a few other simple utilities).

I then issued `composer global require laravel/valet`.

I then issued a `valet install` which went ahead an installed nginx and dnsmasq via homebrew and then reported that valet was successfully installed.

I issued a `valet start` and then tried to ping foo.dev

Ping replied back

`ping: cannot resolve foo.dev: Unknown host`

I restarted the machine, and same thing.

I uninstalled everything, then manually installed nginx and dnsmasq through homebrew, then reinstalled valet.

Same issue.

Checking `brew services list` does not list dnsmasq as an active service, however `ps -ax | grep dnsmasq` display a process that has a time of 0:00.00 and refreshing the ps shows that dnsmasq is constantly restarting (process ID is different on each reload of `ps -ax | grep dnsmasq`).

The same is true for nginx.

**Comment by drbyte** [comment]
What happens if you also run:
```
sudo brew services start dnsmasq
sudo brew services start nginx
```

You may need to configure dnsmasq to do some logging output, and then use that to find out why it's not routing foo.dev for you.

ALSO: what's your Mac set to use for DNS? 
- System Preferences->Network->(Advanced button)->DNS
Should be `127.0.0.1` for dnsmasq to fire properly.

**Comment by indianafollonier** [comment]
The Mac was set to use 192.168.1.1, so changed that to 127.0.0.1, applied that, but still not working. Also, once set to 127.0.0.1 I couldn't access the internet. So I added 192.168.1.1, 8.8.8.8 and 8.8.4.4 (in that order) and I can get out again.

I had tried using `sudo brew services start dnsmasq`and `sudo brew services start nginx`but it made no difference either.

I restart nginx and dnsmasq with sudo with the following results:

```
sudo brew services restart dnsmasq
Stopping `dnsmasq`... (might take a while)
==> Successfully stopped `dnsmasq` (label: homebrew.mxcl.dnsmasq)
==> Successfully started `dnsmasq` (label: homebrew.mxcl.dnsmasq)
sudo brew services restart nginx
Stopping `nginx`... (might take a while)
==> Successfully stopped `nginx` (label: homebrew.mxcl.nginx)
==> Successfully started `nginx` (label: homebrew.mxcl.nginx)
```

I also tried to flush the ins cache using:

`dscacheutil -flushcache && killall -HUP mDNSResponder`

But I got a very odd response:

`No matching processes belonging to you were found`

You suggest that I * may need to configure dnsmasq to do some logging output*. How would I do that?

**Comment by indianafollonier** [comment]
I just tried `dscacheutil -statistics` and got the following `Unable to get details from the cache node.`

**Comment by drbyte** [comment]
I posted a few tips for debugging dnsmasq at https://github.com/laravel/valet/issues/115#issuecomment-241275816

**Comment by indianafollonier** [comment]
How do I know which dnsmasq.conf file is being used? When installing Valet, it creates its own file in `~/.valet/dnsmasq.conf`

Since I had trouble getting it to work, as I mentioned above, I used `brew install dnsmasq nginx` and then `cp /usr/local/opt/dnsmasq/dnsmasq.conf.example /usr/local/etc/dnsmasq.conf`. I then edited /usr/local/etc/dnsmasq.conf` and replaced the default content with `address=/.dev/127.0.0.1` after which I then ran `composer global require laravel/valet`.

Of course, that didn't solve anything.

**Comment by drbyte** [comment]
To enable logging I just added it to `~/.valet/dnsmasq.conf` and restarted dnsmasq. Then it immediately created the logfile I had specified, and I was able to use it to see what problems it was encountering ... or to confirm that it was getting any hits at all when I tried to browse anywhere, particularly to any .dev site.

**Comment by indianafollonier** [comment]
I added the logging info. I figured out which file because valet added `conf-file=/Users/indianafollonier/.valet/dnsmasq.conf` to `/usr/local/etc/dnsmasq.conf`

I then tried to ping to generate a log entry. Ping returned `ping: cannot resolve foo.dev: Unknown host` but nothing was logged (the file waste created). So I assumed the problem was the missing log file, as I touched the file and pinged again, and still, nothing.

**Comment by indianafollonier** [comment]
Hmmm...no idea what I did, but I just copied your "full" `dnsmasq.conf` file, restarted dnsmasq for what seems to be the 10,000th time today, and voila, it works.

**Comment by drbyte** [comment]
As I posted in #115, try using the following as your `~/.valet/dnsmasq.conf` file contents, and `sudo brew services restart dnsmasq`

```
address=/.dev/127.0.0.1

#ignore resolv.conf
no-resolv
no-poll

# listen only on localhost (for dnscrypt)
listen-address=127.0.0.1
# The bind-interfaces directive instructs dnsmasq to bind only to the network interface specified in the listen-address directive, which is important for dnscrypt use
bind-interfaces
# dnscrypt-proxy is on port 53:
server=127.0.0.1#53

# The following directives prevent dnsmasq from forwarding plain names (without any dots) or addresses in the non-routed address space to the parent nameservers.
domain-needed
bogus-priv

# default is 150, or 0 = off
cache-size=0


# Logging
log-facility=/Users/my_user_name/Downloads/dnsmasq.log
log-queries


# fallback servers to use if we can't resolve using our own rules
server=208.67.220.222
server=208.67.220.220
#server=4.4.4.4


#domain-needed This tells dnsmasq to never pass short names to the upstream DNS servers. If the name is not in the local /etc/hosts file then “not found” will be returned.
#bogus-priv All reverse IP (192.168.x.x) lookups that are not found in /etc/hosts will be returned as “no such domain” and not forwarded to the upstream servers.
#no-resolv Do not read resolv.conf to find the servers where to lookup dns.
#no-poll Do not poll resolv.conf for changes
#server=8.8.8.8 Set one or more DNS servers to use when addresses are not local. These are open DNS servers.
#local=/example.com/ Our local domain, queries in these domains are answered from /etc/hosts or the static-hosts files.
#address=/doubleclick.net/127.0.0.1 Use this force an address for the specified domains. e.g to block adverts force doubleclck.net to localhost
#no-hosts This options stops dnsmasq using the local /etc/hosts file as a source for lookups .
#addn-hosts=/etc/dnsmasq_static_hosts.conf Force dnsmasq to use this file for lookups. It is in the same format as /etc/hosts.
#expand_hosts So we can see our local hosts via our home domain without having to repeatedly specify the domain in our /etc/hosts file.
```

What needs to happen is this:
- the mac networking needs to point to the same IP and port that dnsmasq is listening on. By default this is 127.0.0.1:53
- so system preferences needs to point DNS to 127.0.0.1 in order for anything to be directed to dnsmasq
- as long as nothing is interrupting that flow, then all DNS lookup traffic will be sent to dnsmasq for local resolution, and then for any domains it's not configured to serve dnsmasq will defer to 3rd party resolvers as specified by the `server=` entries (You could use 8.8.8.8 or 4.4.4.4 or use the OpenDNS ones I gave above)

But if something in your Mac is interfering with sending DNS traffic to the IP:port that dnsmasq is listening on, then you'll need to figure out what that is and adjust accordingly. 

Like you, I'm a little surprised that a fresh macOS install is having any of these issues. :(

**Comment by drbyte** [comment]
> Hmmm...no idea what I did, but I just copied your "full" dnsmasq.conf file, restarted dnsmasq for what seems to be the 10,000th time today, and voila, it works.

I guess I took too long to type my reply. Glad you got it sorted out ;)

**Comment by indianafollonier** [comment]
I would still like to understand what was causing the issues I was having. I am going to close this issue, however, if you have any way to help me find out what the problems were, please post a reply.

**Comment by drbyte** [comment]
I'd start exploring by removing various lines from the `dnsmasq.conf` file, restarting the service, and testing. See which setting/s stop it from working.

You'll want to leave the `address=` line, and probably all the `server=` lines.

(The `server=` lines let you not have to add those other DNS entries in System Preferences.)

**Comment by mludi** [comment]
Hi @drbyte where can I find the log file?

**Comment by drbyte** [comment]
You have to specify a log file in your `dnsmasq.conf` file. See the discussion above for details. In fact, you could just use the sample file I posted https://github.com/laravel/valet/issues/299#issuecomment-274177318 as a base, and it may "just work".

**Comment by mludi** [comment]
lol after I reinstalled it 4 times, it's working now ... I'll waiting few days and test.

Thanks for helping my @drbyte

**Comment by GeekTNTdotCom** [comment]
I had problems with this too and finally it worked out once I have installed it with sudo
```sudo instal valet```

**Comment by tarranjones** [comment]
Yep @GeekTNTdotCom 's answer fixed it for me (`sudo valet install`).

**Comment by arielestigarribia** [comment]
`sudo install valet` works!

**Comment by flamelin** [comment]
Try to delete  `~/.valet`, `/usr/local/etc/dnsmasq.conf` and reinstall valet.

**Comment by bsesic** [comment]
Besides `*.dev` does not work anymore sind it is a real `TLD`. So use something else like `*.test` or `*.local`.

**Comment by ahmad-shawky** [comment]
This worked for me
```
sudo brew services restart dnsmasq

```

**Comment by ptoly** [comment]
I don't have time to test right now but this has been a periodic headache to me. What I've noticed is that after I run valet fix, the dns.plist will show up on the users folder AS WELL AS the /library/launchitems folder. I see the same with php 5.6. Having startup scripts in both user and system folder is causing FuBar. When I removed all startup scripts from both user and system folders, and run valet install, restart and everything works again...

**Comment by drbyte** [comment]
@ptoly  wrote:
> this has been a periodic headache to me. What I've noticed is that after I run valet fix ... ... causing FuBar. ...

What do you mean by running `valet fix`? That's not a command currently offered by Valet. Where did it come from?

**Comment by ptoly** [comment]
Sorry, my bad. I’m running valet-plus, which does have a fix method...

**Comment by pouyakondori** [comment]
> What happens if you also run:
> 
> ```
> sudo brew services start dnsmasq
> sudo brew services start nginx
> ```
> 
> You may need to configure dnsmasq to do some logging output, and then use that to find out why it's not routing foo.dev for you.
> 
> ALSO: what's your Mac set to use for DNS?
> 
> * System Preferences->Network->(Advanced button)->DNS
>   Should be `127.0.0.1` for dnsmasq to fire properly.

This reply was soooo helpful. The thing which is important is that in Homebrew list "brew services list", there should be a dnsmasq. which the command "brew install dnsmasq" will do it, and then we should start it with "sudo brew services start dnsmasq".
After that I wrote the command "valet start" and everything works fine now.
Thanks a lot again.

**Comment by jamescollins89** [comment]
Try running:
`dnsmasq --test`

If it returns:
`dnsmasq: cannot access directory /Users/mac-user/.config/valet/dnsmasq.d/: No such file or directory`

Copy dnsmasq.d folder from /usr/local/etc to /User/mac-user/.config/valet/dnsmasq.d/ then running `dnsmaqs --test` again returns:
`dnsmasq: syntax check OK`

**Comment by jamiewade** [comment]
I too was in a situation where dnsmasq wasn't working for me. After checking the Advanced network settings for my connection, I noticed that `127.0.0.1` wasn't in the DNS settings, but when I added it, my internet connection stopped working. 

I got around this by adding the default DNS IP address manually, AND adding `127.0.0.1`. So now, all localhost sites resolve when I run `ping testing.app`, and my internet connection works.
